### ISSUE TITLE
UI glitch

### ISSUE THREAD

**Issue by weflower** [gorhill/uMatrix#711]
Sometimes when selecting "Disable/enable matrix filtering for this scope", the UI window doesn't refresh itself correctly:
http://i.imgur.com/07lAKA6.jpg

This is on Chrome version 57.0.2987.37 beta (64-bit). 

I haven't noticed if specific sites are causing the problem. Screen shot above was done on this page:
http://www.feralaudio.com/76-rocklobsterfest-red-lobster-4-sean-clements-hayes-davenport/

If I click the uMatrix toolbar icon to close the menu, then reopen it, the problem is fixed.

**Comment by uBlock-user** [comment]
Yes that's what I have been experiencing since Chromium 57 too - https://github.com/gorhill/uMatrix/issues/689

**Comment by gorhill** [comment]
Duplicate of #689.

Also, quite likely related to this error emitted at the console:

![](https://i.gyazo.com/b1d0f19bae034763505f21c19340558d.png)

I will need some repro steps to be able to investigate.

**Comment by uBlock-user** [comment]
> Install Chromium 57 or 58, install uBO and uMatrix, run them together for a while, click on any cell on the matrix, or enable/disable matrix filtering for the current scope, the cell will be unresponsive and/or UL will glitch and when that happens if you right click and inspect you will see what I posted in the picture.

Can you try that, gorhill ?

**Comment by weflower** [comment]
Thanks, uBlock-user.

gorhill, I also have uBlock Origin 1.11.0 installed. There's no easy steps to guarantee you'll see it, unfortunately.

But the basic steps are:
1. Go to page.
2. Click uMatrix icon on toolbar
3. Click "Disable/enable matrix..."

UI glitches or not at this point.

Not positive yet, but for me if it's going to happen, it seems to happen the first time I'm interacting with a page. If it doesn't glitch the first time, it won't glitch unless I refresh w/out cache.

**Comment by CrisBRM** [comment]
Yesterday I made the switch to chromium (58) and noticed occasionally the popup UI would look normal and be responsive at first, but clicking on a cell would not work until I refreshed the page. I wonder if it's related. Unfortunately, I did not open the console and this does not happen reliably.

**Comment by Gridlocked** [comment]
> Yesterday I made the switch to chromium (58) and noticed occasionally the popup UI would look normal and be responsive at first, but clicking on a cell would not work until I refreshed the page. I wonder if it's related. Unfortunately, I did not open the console and this does not happen reliably.

I'm using Chrome 57 stable on Ubuntu 16.10 and having the same issue though it happens reliably.  I had recently changed 1st party scripts to default to block so I'm editing more frequently.  I have to reload the page in order to make the changes.   I have also seen the UI go whited out similar to the op's report - If I see the UI go white again I'll try and get a screen shot.

**Comment by CrisBRM** [comment]
@Gridlocked yeah, sometimes it turns white.

**Comment by gorhill** [comment]
There is sometimes an error in the javascript of the popup, what I need is a way to reliably reproduce it -- I have witnessed it once, but I don't know how to reproduce it. If I can reproduce at will, I can investigate -- it's probably a trivial fix.

**Comment by uBlock-user** [comment]
Sometimes it happens to me twice or thrice in succession which is so frustrating as I have keep refreshing and hope it goes away 💢

**Comment by migueldemoura** [comment]
Can confirm both issues.

Can reproduce the cell not responding issue - just open a new page and then uMatrix. Clicking the uMatrix icon again fixes it.
Can't reproduce the other one :disappointed:

**Comment by uBlock-user** [comment]
Happens in **0.9.9.4** too however once in a while and not as frequently as in **0.9.3.6**.

**Comment by blockoperation** [comment]
I've managed to capture it a few times in devtools, and it seems that all of the custom properties on the cells (e.g. hostname and reqType) are lost when the poll handler is called. I guess there's some Chromium bug WRT popup contexts or something (I doubt it's an intentional change seeing as it's intermittent).

Modifying uDom to use getAttribute/setAttribute instead of actual properties for DOMList.prop (and then modifying any relevant property accesses in popup.js/settings.js/etc to match) seems to be a viable workaround.

**Comment by Gridlocked** [comment]
I just moved to Fedora from Ubuntu 16.10 and I no longer have to reload to change a cell.

**Comment by gorhill** [comment]
> it seems that all of the custom properties on the cells (e.g. hostname and reqType) are lost when the poll handler is called.

I just reproduced the glitch, and this is what I see happening. I will try to convert to use attributes as suggested.

**Comment by uBlock-user** [comment]
I can reproduce it sporadically, happened twice till now.

**Comment by gorhill** [comment]
What would help is the what the dev console of the extension says, there is no guarantee this is the same issue, it could be a regression from the changes. When this occurs, right click on the panel, click Inspect. This will open the dev console for the popup panel, see what it says.

**Comment by uBlock-user** [comment]
same as this - https://github.com/gorhill/uMatrix/issues/729#issuecomment-286824739

**Comment by gorhill** [comment]
> same as this - #729 (comment)

Then you are not using 0.9.9b5 -- what is reported there does not match at all the source code of 0.9.9b5. Please, be as accurate as possible with information about bug, otherwise this causes huge time-wasting for me.

**Comment by weflower** [comment]
Been testing 0.9.9b5 on two machines for the past few days and haven't seen the issue again. So far so good. Thanks!

**Comment by blockoperation** [comment]
It's still happening for me, but it's very rare.

nodeToExpandosMap gets wiped out when it happens, as if the nodes are being destroyed and recreated behind the scenes (which maybe explains why attributes aren't affected?).

I was in in a hurry when it happened so I couldn't debug any further (and I haven't been able to reproduce it since), but hopefully this info is useful to you.

**Comment by gorhill** [comment]
If this still occur, I would really like the output at the dev console.

**Comment by blockoperation** [comment]
Managed to catch it again. Strangely, nodeToExpandosMap still has a few entries this time. I still have the popup open if you want me to get any more info from it.

![fail](https://cloud.githubusercontent.com/assets/14982921/24075397/7bb260ec-0c12-11e7-9cd0-3c197886c709.png)

**Comment by blockoperation** [comment]
Confirmation that attributes work (I modified the renderMatrixCell functions to set both attributes and expandos for demonstration):

![fail3](https://cloud.githubusercontent.com/assets/14982921/24083343/4305e690-0ccd-11e7-806d-ba99106b33d2.png)

There must be some Chromium bug at work here, surely?

**Comment by gorhill** [comment]
@blockoperation Thanks for investigating. Given the result of your investigation, I will just go ahead and use attributes, this is what uBO does and that probably explain why it does not suffer the issue. It does look like a Chromium bug since v57. The fix for uMatrix should be simple enough, to just modify expandosFromNode to create/lookup a unique attribute and use it to lookup the object holding the expandos.

**Comment by blockoperation** [comment]
Sorry for not responding sooner – your fix works fine here.

It looks like they've finally fixed the upstream bug (assuming this is the same bug: https://bugs.chromium.org/p/chromium/issues/detail?id=702490), and the fix has been backported to stable (no release yet, though).
